pkg/api/grids/grids_test.go

```diff
@@ -0,0 +1,142 @@
+	"github.com/verrazzano/verrazzano-operator/pkg/controller"
+	"github.com/verrazzano/verrazzano-operator/pkg/testutil"
+	"github.com/verrazzano/verrazzano-operator/pkg/testutilcontroller"
+	"github.com/verrazzano/verrazzano-operator/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+var modelBindingPairs = map[string]*types.ModelBindingPair{
+	"test-pair-1": testutil.GetModelBindingPair(),
+}
+
+// TestReturnAllGrids tests that all Grids are returned
+// GIVEN that Grids are configured
+// WHEN I call GetAllGrids
+// THEN all the Grids should be returned
+// AND that the Grids returned should be the correct amount and have valid values
+func TestReturnAllGrids(t *testing.T) {
+	// GIVEN empty model binding pairs, managed clusters and fake kubernetes clients.
+	var clients kubernetes.Interface = k8sfake.NewSimpleClientset()
+	clusters := testutil.GetTestClusters()
+	type args struct {
+		w http.ResponseWriter
+		r *http.Request
+	}
+	tests := []struct {
+		name    string
+		listers controller.Listers
+	}{
+		{
+			name:    "verifyReturnAllGrids",
+			listers: testutilcontroller.NewControllerListers(&clients, clusters, &modelBindingPairs),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var myGrids []Grid
+			Init(tt.listers)
+			resp := httptest.NewRecorder()
+			ReturnAllGrids(resp, httptest.NewRequest("GET", "/grids", nil))
+			assert.Equal(t, http.StatusOK, resp.Code, "expect the http return code to be http.StatusOk")
+			json.NewDecoder(resp.Body).Decode(&myGrids)
+			assert.Len(t, myGrids, 1, "expect returned Grid Array to have 1 entry")
+			assert.Equal(t, "test-coherence", myGrids[0].Name, "expect Grid Name to be test-coherence")
+			assert.Equal(t, "test-coherence-image:latest", myGrids[0].Image, "expect Grid Image Name to be test-coherence-image:latest")
+		})
+	}
+}
+
+// TestReturnSingleGrid tests that the Grid with ID=0 is returned
+// GIVEN that Grid with ID=0 exists
+// WHEN I call ReturnSingleGrid
+// THEN all the Grid with ID=0 should be returned
+// AND that the Grid returned should have the correct ID and have valid values
+func TestReturnSingleGrid(t *testing.T) {
+	// GIVEN empty model binding pairs, managed clusters and fake kubernetes clients.
+	var clients kubernetes.Interface = k8sfake.NewSimpleClientset()
+	clusters := testutil.GetTestClusters()
+	type args struct {
+		w http.ResponseWriter
+		r *http.Request
+	}
+	tests := []struct {
+		name    string
+		listers controller.Listers
+	}{
+		{
+			name:    "verifyReturnSingleGrid",
+			listers: testutilcontroller.NewControllerListers(&clients, clusters, &modelBindingPairs),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			vars := map[string]string{
+				"id": "0",
+			}
+			var myGrid Grid
+			Init(tt.listers)
+			resp := httptest.NewRecorder()
+			req := httptest.NewRequest("GET", "/grids", nil)
```

Comment:
similar to my comment on PR 156 - shouldn't this be /grids/0?

Reply:
No, did that work for you?
We use gorilla mux which looks like it passes the args in 
`	vars := mux.Vars(r)
	key := vars["id"]
`
Thats why, in the test, you see this 
'vars := map[string]string{
				"id": "0",
			}
req = mux.SetURLVars(req, vars)`

I imagine the gorilla muxer does something similar before handing control to the API

Reply:
Yeah it worked for me but I didn't use the recorder directly. I used HandleFunc and passed in the ReturnXXX function as a handler, so the mux might have done it for me.

Reply:
but sorry, yeah I missed the fact you were doing SetURLVars

Reply:
So for example, here's what I did:

```
	request, _ := http.NewRequest("GET", "/applications/0", nil)
	responseRecorder := testutil.InvokeHTTPHandler(request, "/applications/{id}", ReturnSingleApplication)
```